package dataAccess;

import java.io.File;
import java.net.NoRouteToHostException;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.ResourceBundle;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;
import javax.persistence.TypedQuery;

import configuration.ConfigXML;
import configuration.UtilDate;
import domain.Driver;
import domain.Pasajero;
import domain.Ride;
import domain.Solicitud;
import exceptions.AccountAlreadyExistException;
import exceptions.RideAlreadyExistException;
import exceptions.RideMustBeLaterThanTodayException;

/**
 * It implements the data access to the objectDb database
 */
public class DataAccess  {
	private  EntityManager  db;
	private  EntityManagerFactory emf;


	ConfigXML c=ConfigXML.getInstance();

     public DataAccess()  {
		if (c.isDatabaseInitialized()) {
			String fileName=c.getDbFilename();

			File fileToDelete= new File(fileName);
			if(fileToDelete.delete()){
				File fileToDeleteTemp= new File(fileName+"$");
				fileToDeleteTemp.delete();

				  System.out.println("File deleted");
				} else {
				  System.out.println("Operation failed");
				}
		}
		open();
		if  (c.isDatabaseInitialized())initializeDB();
		
		System.out.println("DataAccess created => isDatabaseLocal: "+c.isDatabaseLocal()+" isDatabaseInitialized: "+c.isDatabaseInitialized());

		close();

	}
     
    public DataAccess(EntityManager db) {
    	this.db=db;
    }

	
	
	/**
	 * This is the data access method that initializes the database with some events and questions.
	 * This method is invoked by the business logic (constructor of BLFacadeImplementation) when the option "initialize" is declared in the tag dataBaseOpenMode of resources/config.xml file
	 */	
	public void initializeDB(){
		
		db.getTransaction().begin();

		try {

		   Calendar today = Calendar.getInstance();
		   
		   int month=today.get(Calendar.MONTH);
		   int year=today.get(Calendar.YEAR);
		   if (month==12) { month=1; year+=1;}  
	    
		   
		    //Create drivers 
			Driver driver1=new Driver("driver1@gmail.com","Aitor Fernandez");
			Driver driver2=new Driver("driver2@gmail.com","Ane Gaztañaga");
			Driver driver3=new Driver("driver3@gmail.com","Test driver");

			
			//Create rides
			driver1.addRide("Donostia", "Bilbo", UtilDate.newDate(year,month,15), 4, 7);
			driver1.addRide("Donostia", "Gazteiz", UtilDate.newDate(year,month,6), 4, 8);
			driver1.addRide("Bilbo", "Donostia", UtilDate.newDate(year,month,25), 4, 4);

			driver1.addRide("Donostia", "Iruña", UtilDate.newDate(year,month,7), 4, 8);
			
			driver2.addRide("Donostia", "Bilbo", UtilDate.newDate(year,month,15), 3, 3);
			driver2.addRide("Bilbo", "Donostia", UtilDate.newDate(year,month,25), 2, 5);
			driver2.addRide("Eibar", "Gasteiz", UtilDate.newDate(year,month,6), 2, 5);

			driver3.addRide("Bilbo", "Donostia", UtilDate.newDate(year,month,14), 1, 3);

			
						
			db.persist(driver1);
			db.persist(driver2);
			db.persist(driver3);

	
			db.getTransaction().commit();
			System.out.println("Db initialized");
		}
		catch (Exception e){
			e.printStackTrace();
		}
	}
	
	/**
	 * This method returns all the cities where rides depart 
	 * @return collection of cities
	 */
	public List<String> getDepartCities(){
			TypedQuery<String> query = db.createQuery("SELECT DISTINCT r.from FROM Ride r ORDER BY r.from", String.class);
			List<String> cities = query.getResultList();
			return cities;
		
	}
	/**
	 * This method returns all the arrival destinations, from all rides that depart from a given city  
	 * 
	 * @param from the depart location of a ride
	 * @return all the arrival destinations
	 */
	public List<String> getArrivalCities(String from){
		TypedQuery<String> query = db.createQuery("SELECT DISTINCT r.to FROM Ride r WHERE r.from=?1 ORDER BY r.to",String.class);
		query.setParameter(1, from);
		List<String> arrivingCities = query.getResultList(); 
		return arrivingCities;
		
	}
	/**
	 * This method creates a ride for a driver
	 * 
	 * @param from the origin location of a ride
	 * @param to the destination location of a ride
	 * @param date the date of the ride 
	 * @param nPlaces available seats
	 * @param driverEmail to which ride is added
	 * 
	 * @return the created ride, or null, or an exception
	 * @throws RideMustBeLaterThanTodayException if the ride date is before today 
 	 * @throws RideAlreadyExistException if the same ride already exists for the driver
	 */
	public Ride createRide(String from, String to, Date date, int nPlaces, float price, String driverEmail) throws  RideAlreadyExistException, RideMustBeLaterThanTodayException {
		System.out.println(">> DataAccess: createRide=> from= "+from+" to= "+to+" driver="+driverEmail+" date "+date);
		try {
			if(new Date().compareTo(date)>0) {
				throw new RideMustBeLaterThanTodayException(ResourceBundle.getBundle("Etiquetas").getString("CreateRideGUI.ErrorRideMustBeLaterThanToday"));
			}
			db.getTransaction().begin();
			
			Driver driver = db.find(Driver.class, driverEmail);
			if (driver.doesRideExists(from, to, date)) {
				db.getTransaction().commit();
				throw new RideAlreadyExistException(ResourceBundle.getBundle("Etiquetas").getString("DataAccess.RideAlreadyExist"));
			}
			Ride ride = driver.addRide(from, to, date, nPlaces, price);
			//next instruction can be obviated
			db.persist(driver); 
			db.getTransaction().commit();

			return ride;
		} catch (NullPointerException e) {
			// TODO Auto-generated catch block
			db.getTransaction().commit();
			return null;
		}
		
		
	}
	
	/**
	 * This method retrieves the rides from two locations on a given date 
	 * 
	 * @param from the origin location of a ride
	 * @param to the destination location of a ride
	 * @param date the date of the ride 
	 * @return collection of rides
	 */
	public List<Ride> getRides(String from, String to, Date date) {
		System.out.println(">> DataAccess: getRides=> from= "+from+" to= "+to+" date "+date);

		List<Ride> res = new ArrayList<>();	
		TypedQuery<Ride> query = db.createQuery("SELECT r FROM Ride r WHERE r.from=?1 AND r.to=?2 AND r.date=?3",Ride.class);   
		query.setParameter(1, from);
		query.setParameter(2, to);
		query.setParameter(3, date);
		List<Ride> rides = query.getResultList();
	 	 for (Ride ride:rides){
	 		 if (ride.getnPlaces()>0) {
	 			res.add(ride);
	 		 }
		   
		  }
	 	return res;
	}
	
	public int getAccountType(String email) {//0 not found, 1 Pasajero, 2 Driver
		int accountType=0;
		db.getTransaction().begin();
		Pasajero p=db.find(Driver.class, email);
		if(p==null) {
			p=db.find(Pasajero.class, email);
			if(p!=null) {
				System.out.println(">> DataAccess: getAccountType=> isPasajero");
				accountType=1;
			}
		}
		else {
			System.out.println(">> DataAccess: getAccountType=> isDriver");
			accountType=2;
		}
		db.getTransaction().commit();
		return accountType;
	}
	
	private Pasajero accountExist(String email) {
		Pasajero p=db.find(Pasajero.class, email);
		if(p==null) {
			p=db.find(Driver.class, email);
		}
		return p;
	} 
	
	public Pasajero createPasajero(String email, String password) throws AccountAlreadyExistException { 
		db.getTransaction().begin();
		Pasajero p = accountExist(email);
		if(p!=null) {
			db.getTransaction().commit();
			throw new AccountAlreadyExistException();
		}
		p=new Pasajero(email,password);
		System.out.println(">> DataAccess: createPasajero=> email= "+email+" password= "+password);
		db.persist(p);
		db.getTransaction().commit();
		return p;
	}
	
	public Pasajero getPasajero(String email) {
		System.out.println(">> DataAccess: getPasajero=> email= "+email);
		db.getTransaction().begin();
		Pasajero p=db.find(Pasajero.class, email); 
		db.getTransaction().commit();
		
		return p;		
	}
	
	public Driver createDriver(String email, String password, String name) throws AccountAlreadyExistException { 
		db.getTransaction().begin();
		Driver driver = (Driver)accountExist(email);
		if(driver!=null) { 
			db.getTransaction().commit();
			throw new AccountAlreadyExistException();
		}
		driver=new Driver(email,password,name);
		System.out.println(">> DataAccess: createDriver=> email= "+email+" password= "+password+" name= "+name);
		db.persist(driver);
		db.getTransaction().commit();
		return driver;
	}
	
	public Driver getDriver(String email) {
		System.out.println(">> DataAccess: getDriver=> email= "+email);
		db.getTransaction().begin();
		Driver driver=db.find(Driver.class, email);
		db.getTransaction().commit();
		
		return driver;		
	}
	
	
	public boolean validPassword(String email, String password) {
		db.getTransaction().begin();
		Pasajero p = accountExist(email);
		db.getTransaction().commit();
		if(p==null) {
			System.out.println(">> DataAccess: not valid password");
			return false;
		}
		if(!p.getPassword().equals(password)) {
			System.out.println(">> DataAccess: not valid password");
			return false;
		}
		System.out.println(">> DataAccess: valid password");
		return true;
	}
	/**
	 * This method retrieves from the database the dates a month for which there are events
	 * @param from the origin location of a ride
	 * @param to the destination location of a ride 
	 * @param date of the month for which days with rides want to be retrieved 
	 * @return collection of rides
	 */
	public List<Date> getThisMonthDatesWithRides(String from, String to, Date date) {
		System.out.println(">> DataAccess: getEventsMonth");
		List<Date> res = new ArrayList<>();	
		
		Date firstDayMonthDate= UtilDate.firstDayMonth(date);
		Date lastDayMonthDate= UtilDate.lastDayMonth(date);
				
		
		TypedQuery<Date> query = db.createQuery("SELECT DISTINCT r.date FROM Ride r WHERE r.from=?1 AND r.to=?2 AND r.date BETWEEN ?3 and ?4 AND r.nPlaces>0",Date.class);   
		
		query.setParameter(1, from);
		query.setParameter(2, to);
		query.setParameter(3, firstDayMonthDate);
		query.setParameter(4, lastDayMonthDate);
		
		List<Date> dates = query.getResultList();
	 	 for (Date d:dates){
		   res.add(d);
		  }
	 	return res;
	}
	

	public void open(){
		
		String fileName=c.getDbFilename();
		if (c.isDatabaseLocal()) {
			emf = Persistence.createEntityManagerFactory("objectdb:"+fileName);
			db = emf.createEntityManager();
		} else {
			Map<String, String> properties = new HashMap<>();
			  properties.put("javax.persistence.jdbc.user", c.getUser());
			  properties.put("javax.persistence.jdbc.password", c.getPassword());

			  emf = Persistence.createEntityManagerFactory("objectdb://"+c.getDatabaseNode()+":"+c.getDatabasePort()+"/"+fileName, properties);
			  db = emf.createEntityManager();
    	   }
		System.out.println("DataAccess opened => isDatabaseLocal: "+c.isDatabaseLocal());

		
	}

	public void close(){
		db.close();
		System.out.println("DataAcess closed");
	}
	
	public Pasajero flightBooked(Pasajero c, Solicitud r) {
		open();
		Pasajero p=db.find(Pasajero.class, c);
		ArrayList<Solicitud> lista= new ArrayList<Solicitud>();
		//Solicitud dbSolicitud= db.find(Solicitud.class, r);
		Ride myRide= db.find(Ride.class, r.getRide());
		
		
		if (p==null) {
			System.out.println("Error, el pasajero no esta en la Base de Datos");
		}else {
			lista= p.getSolicitudes();
			lista.add(r);
			db.getTransaction().begin();
			p.setSolicitudes(lista);
			myRide.setBetMinimum(myRide.getnPlaces()-1);
			//dbSolicitud.setRide(myRide);
			db.getTransaction().commit();
			//System.out.println(db.find(Pasajero.class, p).getSolicitudes().getFirst().RideToString());
			
		}
		p= db.find(Pasajero.class, p);
		close();
		return p;
		
	}
	
}
